File: packages/sdk-python/src/stagehand/browser_clipboard.py
from __future__ import annotations

from typing import TYPE_CHECKING, Literal

from ._generated.models import (
    ContextClipboardPasteParams,
    ContextClipboardReadTextResult,
    ContextClipboardTarget,
    ContextClipboardWriteTextParams,
    ContextVoidResult,
    Shortcut,
)
from .rpc_client import RPCClient

if TYPE_CHECKING:
    from .page import Page


class BrowserClipboard:
    def __init__(self, rpc_client: RPCClient) -> None:
        self._rpc_client = rpc_client

    async def read_text(self, *, page: Page | None = None) -> str:
        result = await self._rpc_client.send(
            "context.clipboard_read_text",
            _clipboard_target(page),
            ContextClipboardReadTextResult,
        )
        return result.text

    async def write_text(self, text: str, *, page: Page | None = None) -> None:
        params = ContextClipboardWriteTextParams(text=text)
        if page is not None:
            params.page_id = page.page_id
        await self._rpc_client.send(
            "context.clipboard_write_text",
            params,
            ContextVoidResult,
        )

    async def clear(self, *, page: Page | None = None) -> None:
        await self._rpc_client.send(
            "context.clipboard_clear",
            _clipboard_target(page),
            ContextVoidResult,
        )

    async def paste(
        self,
        *,
        page: Page | None = None,
        shortcut: Shortcut | Literal["ControlOrMeta+V", "Meta+V", "Control+V"] | None = None,
    ) -> None:
        params = ContextClipboardPasteParams()
        if page is not None:
            params.page_id = page.page_id
        if shortcut is not None:
            params = ContextClipboardPasteParams.model_validate({
                **params.model_dump(exclude_unset=True),
                "shortcut": shortcut,
            })
        await self._rpc_client.send(
            "context.clipboard_paste",
            params,
            ContextVoidResult,
        )

    async def copy(self, *, page: Page | None = None) -> None:
        await self._rpc_client.send(
            "context.clipboard_copy",
            _clipboard_target(page),
            ContextVoidResult,
        )

    async def cut(self, *, page: Page | None = None) -> None:
        await self._rpc_client.send(
            "context.clipboard_cut",
            _clipboard_target(page),
            ContextVoidResult,
        )


def _clipboard_target(page: Page | None) -> ContextClipboardTarget:
    return (
        ContextClipboardTarget() if page is None else ContextClipboardTarget(page_id=page.page_id)
    )

File: packages/sdk-ruby/lib/stagehand/browser_clipboard.rb
# frozen_string_literal: true

require_relative "generated/models"
require_relative "rpc_client"

module Stagehand
  # Clipboard surface reached via BrowserContext#clipboard. Port of
  # packages/sdk-python/src/stagehand/browser_clipboard.py. Every method
  # optionally targets a specific page; the active page is used otherwise.
  class BrowserClipboard
    def initialize(rpc_client)
      @rpc_client = rpc_client
    end

    def read_text(page: nil)
      @rpc_client.send("context.clipboard_read_text", target(page), "ContextClipboardReadTextResult")
    end

    def write_text(text, page: nil)
      values = { text: text }
      values[:page_id] = page.page_id unless page.nil?
      @rpc_client.send("context.clipboard_write_text", Models::ContextClipboardWriteTextParams.new(**values), "ContextVoidResult")
      nil
    end

    def clear(page: nil)
      @rpc_client.send("context.clipboard_clear", target(page), "ContextVoidResult")
      nil
    end

    # shortcut: "ControlOrMeta+V" (default), "Meta+V", or "Control+V".
    def paste(page: nil, shortcut: nil)
      values = {}
      values[:page_id] = page.page_id unless page.nil?
      values[:shortcut] = shortcut unless shortcut.nil?
      @rpc_client.send("context.clipboard_paste", Models::ContextClipboardPasteParams.new(**values), "ContextVoidResult")
      nil
    end

    def copy(page: nil)
      @rpc_client.send("context.clipboard_copy", target(page), "ContextVoidResult")
      nil
    end

    def cut(page: nil)
      @rpc_client.send("context.clipboard_cut", target(page), "ContextVoidResult")
      nil
    end

    private

    def target(page)
      page.nil? ? Models::ContextClipboardTarget.new : Models::ContextClipboardTarget.new(page_id: page.page_id)
    end

  end
end
